{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b1f0b226",
   "metadata": {},
   "source": [
    "# ðŸŸ Proyecto OrestiasGen â€“ Pipeline Exploratoria Completa en Colab\n",
    "Este notebook implementa el **pipeline completo exploratorio** descrito en el README, utilizando datos pÃºblicos.\n",
    "Incluye:\n",
    "- InstalaciÃ³n de herramientas\n",
    "- Descarga de datos SRA (4 especies de *Orestias*)\n",
    "- Control de calidad\n",
    "- Alineamiento al genoma de referencia\n",
    "- Llamado de SNPs\n",
    "- Carga del VCF\n",
    "- EstadÃ­stica descriptiva\n",
    "- PCA y visualizaciones\n",
    "\n",
    "âš ï¸ *Este pipeline es exploratorio y opera con datos reales, por lo que puede tardar varios minutos en ejecutarse.*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88857e26",
   "metadata": {},
   "source": [
    "## ðŸ”§ InstalaciÃ³n de herramientas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "291c15ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "!apt-get update\n",
    "!apt-get install -y sra-toolkit\n",
    "!pip install scikit-allel pandas matplotlib seaborn biopython\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9abd9508",
   "metadata": {},
   "source": [
    "## ðŸ” ConfiguraciÃ³n de SRA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f8a0393",
   "metadata": {},
   "outputs": [],
   "source": [
    "!vdb-config --interactive\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "218e4a3d",
   "metadata": {},
   "source": [
    "## ðŸ“ Crear carpetas de trabajo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0cc550f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "folders = [\"data\", \"fastq\", \"qc\", \"bam\", \"vcf\"]\n",
    "for f in folders:\n",
    "    os.makedirs(f, exist_ok=True)\n",
    "\n",
    "folders"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51f34409",
   "metadata": {},
   "source": [
    "## ðŸ“¥ Descarga de datos SRA\n",
    "Se usarÃ¡n accesiones ejemplo de *Orestias* (placeholder; puedes reemplazar por accesiones reales)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8648f7b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "sra_ids = [\n",
    "    \"SRR390728\",  # EJEMPLO â€” reemplazar por Orestias\n",
    "    \"SRR390727\",\n",
    "    \"SRR390726\",\n",
    "    \"SRR390725\"\n",
    "]\n",
    "\n",
    "for sra in sra_ids:\n",
    "    !prefetch {sra}\n",
    "    !fasterq-dump {sra} -O fastq/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "961f9a22",
   "metadata": {},
   "source": [
    "## ðŸ§¼ Control de calidad con FastQC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97a6c2fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "!fastqc fastq/*.fastq -o qc/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cdea17e",
   "metadata": {},
   "source": [
    "## ðŸ§¬ Descargar genoma de referencia (**Orestias ascotanensis**)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80c1e140",
   "metadata": {},
   "outputs": [],
   "source": [
    "!wget -O data/ore_ref.fasta https://ftp.ncbi.nlm.nih.gov/genomes/all/placeholder_reference.fasta\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "641f55ac",
   "metadata": {},
   "source": [
    "## ðŸ§² Indexar el genoma de referencia (BWA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ec1c918",
   "metadata": {},
   "outputs": [],
   "source": [
    "!apt-get install -y bwa samtools bcftools\n",
    "!bwa index data/ore_ref.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "814b511e",
   "metadata": {},
   "source": [
    "## ðŸŽ¯ Alineamiento de FASTQ a referencia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96ade0fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import glob\n",
    "\n",
    "fastqs = glob.glob(\"fastq/*.fastq\")\n",
    "for fq in fastqs:\n",
    "    sample = fq.split(\"/\")[-1].replace(\".fastq\",\"\")\n",
    "    !bwa mem data/ore_ref.fasta {fq} | samtools sort -o bam/{sample}.bam\n",
    "    !samtools index bam/{sample}.bam\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4632bdea",
   "metadata": {},
   "source": [
    "## ðŸ§¬ Llamado de SNPs con bcftools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba7b28e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "!samtools mpileup -uf data/ore_ref.fasta bam/*.bam | bcftools call -mv -Oz -o vcf/orestias_raw.vcf.gz\n",
    "!bcftools index vcf/orestias_raw.vcf.gz\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2aaf1789",
   "metadata": {},
   "source": [
    "## ðŸ” Filtrado bÃ¡sico de SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab203a13",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "!bcftools filter -i 'QUAL>20 && DP>3' vcf/orestias_raw.vcf.gz -Oz -o vcf/orestias_filt.vcf.gz\n",
    "!bcftools index vcf/orestias_filt.vcf.gz\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6c8e4ce",
   "metadata": {},
   "source": [
    "## ðŸ”¬ Cargar VCF y explorar SNPs con scikit-allel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ebd79dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import allel\n",
    "import numpy as np\n",
    "\n",
    "callset = allel.read_vcf(\"vcf/orestias_filt.vcf.gz\")\n",
    "gt = allel.GenotypeArray(callset['calldata/GT'])\n",
    "\n",
    "print(\"N SNPs:\", gt.shape[0])\n",
    "print(\"N muestras:\", gt.shape[1])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "deb815de",
   "metadata": {},
   "source": [
    "## ðŸ“ˆ Frecuencias alÃ©licas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b586b35c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "ac = gt.count_alleles()\n",
    "af = ac.to_frequencies()[:,1]\n",
    "import matplotlib.pyplot as plt\n",
    "plt.hist(af, bins=30)\n",
    "plt.title(\"DistribuciÃ³n de frecuencias alÃ©licas\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ecf6334",
   "metadata": {},
   "source": [
    "## ðŸ§© PCA â€“ estructura genÃ©tica"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9549ae2",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "gn = gt.to_n_alt()\n",
    "gn_filled = np.nan_to_num(gn, nan=0)\n",
    "\n",
    "coords, model = allel.pca(gn_filled, n_components=2)\n",
    "\n",
    "plt.figure(figsize=(6,6))\n",
    "plt.scatter(coords[:,0], coords[:,1])\n",
    "plt.xlabel(\"PC1\")\n",
    "plt.ylabel(\"PC2\")\n",
    "plt.title(\"PCA â€“ Orestias (exploratorio)\")\n",
    "plt.show()\n"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 5
}
